### PR TITLE
Add ldbstore quit channel to terminate running loops

### PR DESCRIPTION
**Pull request by janos**
This PR tries to prevent Swarm stop blocking by terminating running loops, one of which may be a garbageCollect inside writeBatches.

Related issue: https://github.com/ethersphere/go-ethereum/issues/678.

### CODE REVIEW

**Comment by janos** [comment]
@nolash @justelad Could you try to reproduce the bug https://github.com/ethersphere/go-ethereum/issues/678 with this code on your local nodes?

File: swarm/storage/ldbstore.go [hunk 1]
```diff
@@ -549,26 +551,44 @@ func (s *LDBStore) doPut(chunk *Chunk, index *dpaDBIndex, po uint8) {
...
+				done := make(chan struct{})
+				go func() {
+					s.collectGarbage(gcArrayFreeRatio)
+					close(done)
+				}()
+
+				e = s.entryCnt
+				select {
+				case <-s.quit:
+					break mainLoop
```
Comment (acud) [inline swarm/storage/ldbstore.go]: wouldn't this be more elegant to implement using `context.WithCancel`?

**Review by acud — COMMENTED** [review]
@janos generally LGTM.

a small thought on this: I think that `s.quit` might fall into the use case of `context.WithCancel` - i.e. initialize a separate background context for garbage collection, then call the CancelFunc on shutdown (and it would give you also possibility to stop gc from within the context of `collectGarbage`)

File: swarm/storage/ldbstore.go [hunk 2]
```diff
@@ -549,26 +551,44 @@ func (s *LDBStore) doPut(chunk *Chunk, index *dpaDBIndex, po uint8) {
...
+			s.lock.Lock()
+			b := s.batch
+			e := s.entryCnt
+			d := s.dataIdx
+			a := s.accessCnt
+			c := s.batchC
+			s.batchC = make(chan bool)
+			s.batch = new(leveldb.Batch)
+			err := s.writeBatch(b, e, d, a)
+			// TODO: set this error on the batch, then tell the chunk
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/storage/ldbstore.go:569]: well it is part of the chunk refactor code

**Review by zelig — APPROVED** [review]

File: swarm/storage/ldbstore.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (acud) [reply swarm/storage/ldbstore.go:569]: 👍

**Review by acud — APPROVED** [review]

File: swarm/storage/ldbstore.go (hunk 1 above)
Comment (janos) [reply swarm/storage/ldbstore.go]: I think that a single channel is the simplest for of cancelation in this case. We would have to check ctx.Done, the same way as it is done here with s.quit, but with checking ctx.Err and a bit more code to construct context and pass it around.

**Review by zelig — APPROVED** [review]

**Pull request merged**